sql_grader/models.py

```diff
@@ -0,0 +1,222 @@
+    submission_query = None
+    submission_result = None
+
+    def __init__(
+            self,
+            database,
+            answer_query,
+            verify_query=None,
+            is_ordered=True,
+    ):
+        self.database = database
+        self.is_ordered = is_ordered
+        self.answer_query = answer_query
+        self.verify_query = verify_query
+        self.answer_result, _ = SqlProblem.run_query(
+            database,
+            answer_query,
+            verify_query,
+        )
+
+    def attempt(self, query):
+        self.submission_query = query
+        submission_result, error = SqlProblem.run_query(
+            self.database,
+            query,
+            self.verify_query,
+        )
+        self.submission_result = submission_result
+        # print('submission_result', submission_result)
+        comparison = SqlProblem.compare_rows(
+            self.answer_result,
+            self.submission_result,
+            is_ordered=self.is_ordered,
+        )
+        # print('comparison', comparison)
+        return comparison, error
+
+    @classmethod
+    def create_database(cls, path_sql):
+        """
+        Create a new in-memory database, initialized via SQL script
+
+        This only needs run once, at startup.
+        """
+        with open(path_sql, 'r') as sql:
+            query = sql.read()
+        connection = cls.create_database_from_sql(query)
+        return connection
+
+    @classmethod
+    def create_database_from_sql(cls, query):
+        connection = sqlite3.connect(':memory:', check_same_thread=False)
+        with connection:
+            connection.executescript(query)
+        return connection
+
+    @staticmethod
+    def clone_database(source):
+        """
+        Copy the contents of a source database into a new in-memory database
+
+        This should be run for each request.
+        """
+        destination = sqlite3.connect(':memory:', check_same_thread=False)
+        query = ''.join(line for line in source.iterdump())
+        destination.executescript(query)
+        return destination
+
+    @classmethod
+    def run_query(cls, source, query, verify_query=None):
+        """
+        Execute the provided SQL query against a copy of the database
+        """
+        def run(database, query):
+            result = []
+            message = None
+            with database as connection:
+                try:
+                    for row in connection.execute(query):
+                        # print('row', row)
+                        result.append(row)
+                except Exception as error:
+                    print('ERROR', error)
+                    result = None
+                    message = unicode(error)
+                    # TODO: more robust error handling
+            return result, message
+        database = cls.clone_database(source)
+        # print('*' * 20)
+        result, error = run(database, query)
+        if verify_query:
```

Comment:
Queries that modify the state of the DB require an additional verification step.